Tests/AsyncQueueTests/SemaphoreTests.swift

```diff
@@ -0,0 +1,91 @@
+// MIT License
+//
+// Copyright (c) 2022 Dan Federman
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import XCTest
+
+@testable import AsyncQueue
+
+final class SemaphoreTests: XCTestCase {
+
+    // MARK: XCTestCase
+
+    override func setUp() async throws {
+        try await super.setUp()
+
+        systemUnderTest = Semaphore()
+    }
+
+    override func tearDown() async throws {
+        try await super.tearDown()
+
+        while await systemUnderTest.isWaiting {
+            await systemUnderTest.signal()
+        }
+    }
+
+    // MARK: Behavior Tests
+
+    func test_wait_suspendsUntilEqualNumberOfSignalCalls() async {
+        let counter = Counter()
+        let iterationCount = 1_000
+
+        var waits = [Task<Void, Never>]()
+        for _ in 1...iterationCount {
+            waits.append(Task {
+                await self.systemUnderTest.wait()
+                await counter.increment()
+            })
+        }
+
+        var signals = [Task<Void, Never>]()
+        for _ in 0..<(iterationCount-1) {
+            signals.append(Task {
+                await self.systemUnderTest.signal()
+                let count = await counter.count
+                XCTAssertEqual(0, count)
+            })
+        }
+
+        for signal in signals {
+            await signal.value
+        }
+
+        await self.systemUnderTest.signal()
+
+        for wait in waits {
+            await wait.value
+        }
```

Comment:
We want to execute the waits before the signals right? In the wait is where we increment. It seems like we need to execute the waits first for L64 (`XCTAssertEqual(0, count)`) to be a meaningful test.

Reply:
Indeed! This one is a little tricky and I don't love it.

We can't `await wait.value` until we've signaled the semaphore or we'll deadlock. So what I do is spawn 1000 `Task`s that wait for the semaphore assuming that at least one of them will execute before we finish waiting for the signals. Based on code coverage when run locally, this does occur.

I'm open to better ways to write this test, but Swift's concurrency model doesn't give me a great way to say "wait until this task has started" or "wait until this task has suspended", which is what we'd need to make this test more deterministically meaningful.

Reply:
> We can't `await wait.value` until we've signaled the semaphore or we'll deadlock.

👍 

> I'm open to better ways to write this test, but Swift's concurrency model doesn't give me a great way to say "wait until this task has started" or "wait until this task has suspended", which is what we'd need to make this test more deterministically meaningful.

I got into the code and I see what you are saying. The way this test is written I believe that it is _possible_ for it to fail intermittently. I can see the test failing if the signaling tasks were scheduled more frequently than the waiting tasks. That could result in a situation where the counter was incremented before all signaling was complete.

I also am not clear on how to make this test more deterministic 🤔

Reply:
> I also am not clear on how to make this test more deterministic 🤔

This is why I am running all tests 100 times in CI, and using large numbers of repetitions in tests. I'm doing my best to make tests deterministic, but some tests aren't possible to make so. Therefore I rely on repetition to catch issues.

Locally as I develop I run tests 1000 times after I've made a change I'm not sure about just to make sure 😅

Clearly not ideal, but these are the tools I see so I'm using them!

Reply:
Thanks for talking this through!

Reply:
Re-closing the loop here: I pulled this test out into https://github.com/dfed/swift-async-queue/pull/5, and was able to get it to deterministically fail when the we `signal` too many times!